### ISSUE TITLE
Axes3D quiver: variable length arrows

### ISSUE THREAD

**Issue by khs26** [matplotlib/matplotlib#4211]
Would there be interest in allowing the `length` argument to `Axes3d.quiver()` optionally be an array instead of just a single float?

It's a feature I'd like, but I wasn't sure whether to do it properly and make a PR or if there's a good reason for keeping things as is.

**Comment by tacaswell** [comment]
You can also specify the `(u, v, w)` components.  I think specifying `length` to be a vector as well would be degenerate and cause confusion.

**Comment by khs26** [comment]
Sure, but in that case, shouldn't the normalisation of the `u, v, w` components be disabled?

I haven't really use masked numpy arrays, but I think the snippet below from `mplot3d.py` guarantees that the `u, v, w` components always represent a unit vector:

```
        # Normalize rows of UVW
        # Note: with numpy 1.9+, could use np.linalg.norm(UVW, axis=1)
        norm = np.sqrt(np.sum(UVW**2, axis=1))

        # If any row of UVW is all zeros, don't make a quiver for it
        mask = norm > 1e-10
        XYZ = XYZ[mask]
        UVW = UVW[mask] / norm[mask].reshape((-1, 1))
```

**Comment by khs26** [comment]
Just to confirm, if I turn off the normalisation, things do seem to work as I'd hoped. I don't know whether there would be undesired knock-on effects from this change though.

**Comment by tacaswell** [comment]
That does seem like what it _should_ be doing (as that is how 2D quiver works).

**Comment by tacaswell** [comment]
Can you put in a PR making that change?

**Comment by khs26** [comment]
Sure. I'll make one and reference this issue. I'm working on something else at the moment, but should get round to it in a few hours.

**Comment by tacaswell** [comment]
Thanks!

Don't stress too much about getting it done right away the deadline for this to make it into the next version is July (you also, don't procrastinate _too_ long).

**Comment by khs26** [comment]
Another small thing, where should the arrows be?

At the moment, the arrows seem to point at `x, y, z` from `x-u, y-v, z-w`. This also seems counterintuitive. I guess there are three sensible options:

1) Leave it as is.
2) Have the arrows point from `x, y, z` to `x+u, y+v, z+w` so that the arrows start at the specified `x, y, z` coordinates.
3) Centre the arrows about `x, y, z`, so they point from `x-u/2, y-v/2, z-w/2` to `x+u/2, y+v/2, z+w/2`.

My feeling is that option 3 is the most intuitive of the three.

**Comment by khs26** [comment]
Here is an image demonstrating the current state of things. It's a representation of a molecule, so the green lines are bonds between atoms on the `x, y, z` coordinates made using `axes3d.plot`. The arrows are from `axes3d.quiver`.

![Illustrative example prior to fixing normalisation](http://i.stack.imgur.com/uZYJv.png)

**Comment by WeatherGod** [comment]
The documentation states that the xyz coordinates are the default locations
for the tip of the arrows. You can use the "pivot" keyword argument to
change that. That feature might only be in the master branch though.

**Comment by khs26** [comment]
The `pivot` keyword for the 2d quiver plots seems to default to `tail`. Shouldn't they ideally be consistent? I would have thought that having the API call be essentially the same for 2d and 3d quiver, just with two extra sets of coordinates.

**Comment by tacaswell** [comment]
Yes, they should be consistent.

Despite my normal insistence on maintaining the API, I think this is an ok
default to change as the 3D quiver is new and it _should_ match the 2D
defaults.

**Comment by DanHickstein** [comment]
I agree that the normalization should be removed (or maybe made into a keyword argument defaulting to False?). Also, pivoting from the tail seems like the most sensible to match 2D quiver. It seems like this never got changed...

Should I put in a PR for this?

**Comment by WeatherGod** [comment]
The sooner, the better, as the 2.0 release is getting put together right
now. 2.0 will allow for changes in default settings.

**Comment by DanHickstein** [comment]
Okay, PR created. Hopefully I did it correctly. I'm not too GitHub savvy. :)

**Comment by mdboom** [comment]
@WeatherGod: As we've been making default changes, for many we've been adding a method to get the old behavior (usually just through `matplotlib.style.use('classic')`).  How important do you think it is to do that for this as well?  I think for things of fairly low usage, it's a little more ok to "just break" and not have a safety valve to get the old behavior.

**Comment by DanHickstein** [comment]
It would be easy enough to include an option for the normalization. Though it's hard for me to think of a scenario where someone would desire this behavior.

**Comment by tacaswell** [comment]
The use case is when you are only interested in the direction, not the
magnitude. I can imagine a case where the size variation is big enough that
you can not see both the biggest and smallest vectors.

**Comment by DanHickstein** [comment]
Yes, I can see that there are some circumstances where the user would want to normalize their data, but should this be built into the quiver function? It's not in the 2D case...

**Comment by WeatherGod** [comment]
@mdboom, +1 on having an option to normalize since the feature already
exists, +0 on not bothering to add an rcParam so that the "classic" style
would pick up the normalization behavior.

**Comment by DanHickstein** [comment]
Sounds good. I updated the PR to include a "normalize" keyword argument that defaults to False and controls the behavior.

**Comment by WeatherGod** [comment]
Closed by #5458
